src/source/schematron.c

```diff
@@ -420,6 +420,170 @@ static bool _req_src_236_2(xmlXPathContextPtr context)
 	return res;
 }
 
+
+static bool _req_src_346_1_sub5(xmlNodePtr component, xmlXPathContextPtr context, const char *name)
+{
+	/* .//oval-def:definition[@id eq $n/@name]) */
+	bool exists_oval = false;
+	char *oval_xpath = oscap_sprintf(".//oval-def:definition[@id='%s']", name);
+	xmlXPathObjectPtr oval_definitions = xmlXPathNodeEval(component, BAD_CAST oval_xpath, context);
+	for (int m = 0; m < oval_definitions->nodesetval->nodeNr; m++) {
+		xmlNodePtr oval_definition = oval_definitions->nodesetval->nodeTab[m];
+		char *oval_def_id = (char *) xmlGetProp(oval_definition, BAD_CAST "id");
+		dD("Found OVAL definition id='%s'", oval_def_id);
+		free(oval_def_id);
+		exists_oval = true;
+	}
+	free(oval_xpath);
+	xmlXPathFreeObject(oval_definitions);
+	return exists_oval;
+}
+
+static bool _req_src_346_1_sub4(xmlNodePtr check_content_ref_node, xmlNodePtr catalog, xmlXPathContextPtr context)
+{
+
+	char *href = (char *) xmlGetProp(check_content_ref_node, BAD_CAST "href");
+	/* xcf:get-component(xcf:get-component-ref($m/cat:catalog, $o/@href) */
+	xmlNodePtr component_ref =  _xcf_get_component_ref(catalog, href, context);
+	free(href);
+	if (component_ref == NULL) {
+		dD("component_ref not found");
+		return false;
+	}
+	xmlNodePtr component = _xcf_get_component(component_ref, context);
+	if (component == NULL) {
+		dD("component not found\n");
+		return false;
+	}
+
+	/* not(exists($n/@name)) or exists(xcf:get-component(xcf:get-component-ref($m/cat:catalog, $n/@href))//oval-def:definition[@id eq $n/@name]) */
+	char *name = (char *) xmlGetProp(check_content_ref_node, BAD_CAST "name");
+	bool res = true;
+	if(name != NULL) {
+		res = _req_src_346_1_sub5(component, context, name);
+	}
+
+	free(name);
+	return res;
+}
+
+static bool _req_src_346_1_sub3(xmlNodePtr rule_node, xmlNodePtr catalog, xmlXPathContextPtr context)
+{
+	bool res = true;
+	char *rule_id = (char *) xmlGetProp(rule_node, BAD_CAST "id");
+
+	/* note: $n is an xccdf:Rule */
+	/* xccdf:check[@system eq 'http://oval.mitre.org/XMLSchema/oval-definitions-5']//xccdf:check-content-ref[not(xcf:is-external-ref($m/cat:catalog, @href) cast as xsd:boolean)] */
+	xmlXPathObjectPtr check_content_refs = xmlXPathNodeEval(rule_node, BAD_CAST "xccdf:check[@system='http://oval.mitre.org/XMLSchema/oval-definitions-5']/xccdf:check-content-ref[@name]", context);
```

Comment:
This probably requires a comment or a change because it isn't obvious.

The original sub-expression in the comment above `xccdf:check[@system eq 'http://oval.mitre.org/XMLSchema/oval-definitions-5']//xccdf:check-content-ref[not(xcf:is-external-ref($m/cat:catalog, @href) cast as xsd:boolean)]` doesn't contain `@name` attribute, yet it appears here.

It seems to be correct, later in the original expression, the `xccdf:check-content-ref` elements are checked for the presence of `@name` attribute and the elements without it pass the check.


I suggest matching all `xccdf:check-content-ref` elements in this expression and move the `@name` check into the `_req_src_346_1_sub4`. That way we can be more close to the original code.

```suggestion
	xmlXPathObjectPtr check_content_refs = xmlXPathNodeEval(rule_node, BAD_CAST "xccdf:check[@system='http://oval.mitre.org/XMLSchema/oval-definitions-5']/xccdf:check-content-ref", context);
```

Another option is to add a comment explaining this difference.

Reply:
It makes sense. I'll accept the suggestion.